discovery-provider/src/tasks/user_replica_set.py

```diff
@@ -0,0 +1,267 @@
+        logger.info(f"user_replica_set.py | Content Node Processing Adding {value_obj['content_node']}")
+        invalidate_old_cnode_record(session, content_node_id)
+        session.add(value_obj["content_node"])
+
+    return num_total_changes, user_ids
+
+# Reconstruct endpoint string from primary and secondary IDs
+# Note that this is BEST EFFORT and may fail, however unlikely
+# In the case of failure where a given user has valid primary/secondaries but no
+# endpoint string, a client must fetch the endpoint associated with ID
+# Indexing CANNOT block on this endpoint string optimization.
+def get_endpoint_string_from_sp_ids(
+        self,
+        update_task,
+        primary,
+        secondaries,
+        redis
+):
+    sp_factory_inst = None
+    endpoint_string = None
+    primary_endpoint = None
+    try:
+        # Get primary cache key
+        primary_cache_key = get_sp_id_key(primary)
+        # Attempt to fetch from cache
+        primary_info_cached = get_pickled_key(update_task.redis, primary_cache_key)
+        if primary_info_cached:
+            primary_endpoint = primary_info_cached[1]
+            logger.info(f"user_replica_set.py | CACHE HIT FOR {primary_cache_key}, found {primary_info_cached}")
+
+        # If cache miss, fetch from eth-contract
+        if not primary_endpoint:
+            logger.info(f"user_replica_set.py | CACHE MISS FOR {primary_cache_key}, found {primary_info_cached}")
+            if sp_factory_inst is None:
+                sp_factory_inst = get_sp_factory_inst(self, update_task)
+            primary_cn_endpoint_info = sp_factory_inst.functions.getServiceEndpointInfo(
+                content_node_service_type,
+                primary
+            ).call()
+            logger.info(primary_cn_endpoint_info)
+            primary_endpoint = primary_cn_endpoint_info[1]
+
+        endpoint_string = "{}".format(primary_endpoint)
+        for secondary_id in secondaries:
+            secondary_endpoint = None
+            secondary_cache_key = get_sp_id_key(secondary_id)
+            secondary_info_cached = get_pickled_key(update_task.redis, secondary_cache_key)
+            if secondary_info_cached:
+                secondary_endpoint = secondary_info_cached[1]
+                logger.info(f"user_replica_set.py | CACHE HIT FOR {secondary_cache_key}, found {secondary_info_cached}")
+
+            if not secondary_endpoint:
+                logger.info(
+                    f"user_replica_set.py | CACHE MISS FOR {secondary_cache_key}, found {secondary_info_cached}"
+                )
+                if sp_factory_inst is None:
+                    sp_factory_inst = get_sp_factory_inst(self, update_task)
+                secondary_info = sp_factory_inst.functions.getServiceEndpointInfo(
+                    content_node_service_type,
+                    secondary_id
+                ).call()
+                secondary_endpoint = secondary_info[1]
```

Comment:
this block of code is repeated from lines 154-162. could instead make an array `spIDs = [primary, ...secondaries]` and then call this once in a for loop over that

Reply:
i think that actually makes the code worse - I considered your suggestion but decided to explicitly separate the logic for primary vs secondaries. i do agree and think there is a better abstraction though to avoid code repetition, will update

Reply:
fixed